File: linkedlist/reverselinkedlist.js

function listNode(val){
  this.val = val;
  this.next = null;
}

  var head = new listNode(10)
  head.next = new listNode(20) 
  head.next.next = new listNode(30) 

function reverseList (head){
  if((!head) || (!head.next)){
    return head
  } 
  var runner = head;
  var prev = null
  while(runner){
    var temp = runner.next;
    runner.next = prev;
    prev = runner
    runner = temp;


  }
  return prev
}

  console.log(reverseList (head))
  // console.log(reverseList (head))

// time complexity o(N) and space complexity o(n)

// function reverseList(){
//   if(head == null) {
//     return 
//   }
//   var stack = []
//   temp = head;
//   while(temp != null){
//     stack.push(temp);
//     temp = temp.next;
//   }
//   temp = stack.top();
//   // temp = stack[stack.length-1]
//   head = temp;
//   stack.pop();

//   while(!stack.empty()){
//     temp.next = stack.top();
//     stack.pop();
//     temp = temp.next;
//   }
//   temp.next = null
// }

File: linkedlist/reverselinkedlist.py
class listNode:
  def __init__(self,val):
    self.val = val
    self.next = None

head =  listNode(10)
head.next = listNode(20) 
head.next.next = listNode(30) 



class LinkedList:
  def __init__(self):
    self.head = None

  def reverselist(self, head):
    if((not head) or (not head.next)):
      return head
    
    runner = head
    prev = None

    while(runner):
      temp = runner.next
      runner.next = prev
      prev = runner
      runner = temp
    self.head = prev

  def printLinkedList(self):
    temp = self.head
    while(temp):
      print(temp.val)
      temp = temp.next

if __name__ == "__main__":
  print("welcome to python")
  list = LinkedList()
  print(list.reverselist(head))
  list.printLinkedList()